notebooks/k_nearest_neighbors.livemd

```diff
@@ -0,0 +1,833 @@
+  KFold.apply_k_fold(x, y, param_space_size: param_space_size, num_splits: k)
+
+results =
+  DF.new(accuracy: res, num_neighbors: Nx.add(Nx.iota({param_space_size}), min_num_neighbors))
+```
+
+To potentially improve our KNN model's performance, we can try setting the weights parameter to `:distance` in `KFold.one_round_fold()`. Next, we can visualize the results of KFold using a plot.
+
+```elixir
+Vl.new(
+  title: [
+    text: "Accuracy vs Number of Neighbors",
+    offset: 20
+  ],
+  width: 630,
+  height: 630
+)
+|> Vl.data_from_values(results)
+|> Vl.mark(:line, color: :gray, stroke_dash: [5, 5], point: true)
+|> Vl.encode_field(:x, "num_neighbors", type: :quantitative, scale: [domain: [1, 31]])
+|> Vl.encode_field(:y, "accuracy", type: :quantitative, scale: [domain: [0.7, 0.85]])
+```
+
+```elixir
+best_neighbors = Nx.add(Nx.argmax(res), min_num_neighbors) |> Nx.to_number()
+```
+
+Ok, let's test our best model on test data.
+
+```elixir
+best_model =
+  KNearestNeighbors.fit(train_data, train_labels, num_neighbors: best_neighbors, num_classes: 3)
+
+final_prediction = KNearestNeighbors.predict(best_model, test_data)
+Scholar.Metrics.accuracy(test_labels, final_prediction)
+```
+
+Pretty well, now let's move on to a regression task using KNN.
+
+<!-- livebook:{"branch_parent_index":1} -->
+
+## Regression
+
+KNNs can also be used in regression tasks. We will explore this on a dataset about Airfoil Self-Noise. Here's a brief description of the dataset:
+
+**Data Set Information**:
+
+The NASA data set comprises different size NACA 0012 airfoils at various wind tunnel speeds and angles of attack. The span of the airfoil and the observer position were the same in all of the experiments.
+
+**Attribute Information**:
+
+This problem has the following inputs:
+
+1. Frequency, in Hertzs.
+2. Angle of attack, in degrees.
+3. Chord length, in meters.
+4. Free-stream velocity, in meters per second.
+5. Suction side displacement thickness, in meters.
+
+The only output is:
+
+1. Scaled sound pressure level, in decibels.
+
+Dua, D. and Graff, C. (2019). UCI Machine Learning Repository [http://archive.ics.uci.edu/ml]. Irvine, CA: University of California, School of Information and Computer Science.
+
+```elixir
+dataset =
+  Req.get!(
+    "https://archive.ics.uci.edu/ml/machine-learning-databases/00291/airfoil_self_noise.dat"
+  ).body
+
+column_names = [
+  "frequency",
+  "angle_of_attack",
+  "chord_length",
+  "free_stream_velocity",
+  "suction_side_displacement_thickness",
+  "scaled_sound_pressure_level"
+]
+
+dataset_df = DF.load_csv!(dataset, header: false, delimiter: "\t") |> DF.rename(column_names)
+labels = DF.select(dataset_df, "scaled_sound_pressure_level")
+data = DF.discard(dataset_df, "scaled_sound_pressure_level")
+```
+
+Below are the KDE plots of the dataset's features. We will analyze them in the next step.
+
+```elixir
+Vl.new(
+  title: [
+    text: "KDE plots of Features",
+    offset: 20
+  ]
+)
```

Comment:
I will simplify the code of KDE plots, but to be honest, I don't have any idea right now.